src/bsp/src/posix/src/Factory.cpp

```diff
@@ -0,0 +1,17 @@
+#include "bsp/Factory.h"
+#include "BSP.h"
+
+#define ROS_SPIN_LOOP_RATE_MS 200
+
+namespace BspFactory {
+
+std::shared_ptr<IBSP> getBSP() {
+    static BSP s_bsp;
+    static bool s_isCreated = false;
+    if (!s_isCreated) {
+        static ros::NodeHandle s_handle("~/");
+        new (&s_bsp) BSP(s_handle, ROS_SPIN_LOOP_RATE_MS);
+    }
```

Comment:
You shouldn't create an object with a new and assign it to a static object. Your initial static object will still be in memory and unused.

You can initialize static variable, the initialization will only be run once and avoid the use of new for the embedded platform

```
    static ros::NodeHandle s_handle("~/");
    static BSP s_bsp(s_handle, ROS_SPIN_LOOP_RATE_MS);

   return &s_bsp
```

Reply:
@xgroleau This is a placement new https://www.geeksforgeeks.org/placement-new-operator-cpp/. It creates the object inside of the memory used by the static object so no heap.

Reply:
The use of this is to have the object allocated first and then get the different dependencies and providing them to the object once with the placement new.

Reply:
But would the proposed initialization worked instead? removing the need of an empty constructor?

Reply:
It would work and would be simple in that case. Placement news might be useful in the future for more complex objects

Reply:
I'm fine with the placement new if the need arise we can use them, but you can probably still change for the proposed alternative since it removes the need of the empty constructor.